patches/src/main/kotlin/app/revanced/patches/spotify/misc/UnlockPremiumPatch.kt

```diff
@@ -27,23 +34,75 @@ val unlockPremiumPatch = bytecodePatch(
-        val attributesMapRegister = 0
-        val instantiateUnmodifiableMapIndex = 1
-        productStateProtoFingerprint.method.addInstruction(
-            instantiateUnmodifiableMapIndex,
-            "invoke-static { v$attributesMapRegister }," +
-                "$EXTENSION_CLASS_DESCRIPTOR->overrideAttribute(Ljava/util/Map;)V",
-        )
+        with(productStateProtoFingerprint.method) {
+            val getAttributesMapIndex = indexOfFirstInstructionOrThrow(Opcode.IGET_OBJECT)
+            val attributesMapRegister = getInstruction<TwoRegisterInstruction>(getAttributesMapIndex).registerA
+
+            addInstruction(
+                getAttributesMapIndex + 1,
+                "invoke-static { v$attributesMapRegister }, " +
+                        "$EXTENSION_CLASS_DESCRIPTOR->overrideAttribute(Ljava/util/Map;)V"
+            )
+        }
 
         // Add the query parameter trackRows to show popular tracks in the artist page.
-        val addQueryParameterIndex = buildQueryParametersFingerprint.stringMatches!!.first().index - 1
-        buildQueryParametersFingerprint.method.replaceInstruction(addQueryParameterIndex, "nop")
+        with(buildQueryParametersFingerprint) {
+            val addQueryParameterConditionIndex = method.indexOfFirstInstructionReversedOrThrow(
+                stringMatches!!.first().index, Opcode.IF_EQZ
+            )
+            method.replaceInstruction(addQueryParameterConditionIndex, "nop")
+        }
 
         // Disable the "Spotify Premium" upsell experiment in context menus.
         with(contextMenuExperimentsFingerprint) {
-            val moveIsEnabledIndex = stringMatches!!.first().index + 2
+            val moveIsEnabledIndex = method.indexOfFirstInstructionOrThrow(
+                stringMatches!!.first().index, Opcode.MOVE_RESULT
+            )
             val isUpsellEnabledRegister = method.getInstruction<OneRegisterInstruction>(moveIsEnabledIndex).registerA
             method.replaceInstruction(moveIsEnabledIndex, "const/4 v$isUpsellEnabledRegister, 0")
         }
+
+        // Make featureTypeCase_ accessible so we can check the home section type in the extension.
+        homeSectionFingerprint.classDef.fields.first { it.name == "featureTypeCase_" }.apply {
+            accessFlags = accessFlags.or(AccessFlags.PUBLIC.value).and(AccessFlags.PRIVATE.value.inv())
+        }
+
+        val protobufListClassName = with(protobufListsFingerprint.originalMethod) {
+            val emptyProtobufListGetIndex = indexOfFirstInstructionOrThrow(Opcode.SGET_OBJECT)
+            getInstruction(emptyProtobufListGetIndex).getReference<FieldReference>()!!.definingClass
+        }
+
+        val protobufListRemoveFingerprint = fingerprint {
+            custom { method, classDef ->
+                method.name == "remove" && classDef.type == protobufListClassName
+            }
+        }
+
+        // Need to allow mutation of the list so the home ad sections can be removed.
```

Comment:
```suggestion
        // Need to allow mutation of the list so the home ads sections can be removed.
```

Other places mention ads in plural, so might as well do here too